### DISCUSSION TITLE
How to capture the exact versions of software running on iCub robot?

### DISCUSSION THREAD

**Discussion by traversaro** [robotology/community#450] [Q&A] [answered]
For **[scientific reproducibility](https://www.osti.gov/servlets/purl/1525948)** or to **provide more information** when I am reporting a problem on a robot, I would like to know what is the suggested way of **dumping exactly the specific version the software** that is running on an iCub robot. There is any specific procedure to follow?

**Comment by traversaro** [comment, marked as answer]
Thanks for the question! 

First of all, for actual deployment of software on the iCub robot, I suggest the [Robot Bazaar](https://robot-bazaar.iit.it/), the software infrastructure that uses Docker to deploy easily complex demos on the iCub robot cluster. However, if you are deploying your own software manually on the iCub cluster, to capture exactly the version of the software running on a machine related to the iCub setup (including the `icub-head` or `pc104` system), please see the following guide. 

## apt packages 

If you are using a Debian or Ubuntu machine, you can easily dump the exact version of all the apt packages installed with the following command:
~~~
apt list --installed > installed_apt_packages.txt
~~~
this will create a file named `installed_apt_packages.txt` that you can attach were you prefer. The content of the file will be lines of this kind: 
~~~
traversaro@IITICUBLAP102:~$ head installed_apt_packages.txt
Listing...
accountsservice/focal-updates,focal-security,now 0.6.55-0ubuntu12~20.04.4 amd64 [installed,automatic]
acl/focal,now 2.2.53-6 amd64 [installed,automatic]
adduser/focal,now 3.118ubuntu2 all [installed,automatic]
adwaita-icon-theme/focal-updates,now 3.36.1-2ubuntu0.20.04.2 all [installed,automatic]
aglfn/focal,now 1.7+git20191031.4036a9c-2 all [installed,automatic]
alsa-topology-conf/focal,now 1.2.2-1 all [installed,automatic]
alsa-ucm-conf/focal-updates,now 1.2.2-1ubuntu0.4 all [installed,automatic]
alsa-utils/focal-updates,now 1.2.2-1ubuntu2 amd64 [installed,automatic]
apg/focal,now 2.2.3.dfsg.1-5 amd64 [installed,automatic]
~~~

## git repos 

If you are using some software that is compiled from git repositories, as for example when using the robotology-superbuild from source, you may want to easily capture the exact version of the repo are you using. For doing so, **the first step is to commit any uncommitted file** and **push (even on a fork) any not pushed commit**. Unless you do this, it is impossible in the future to know the content of the locally modified files. 

After you have done that, install the [`vcstool`](https://github.com/dirk-thomas/vcstool) tool following the official instructions: https://github.com/dirk-thomas/vcstool#how-to-install-vcstool (note: if you are using conda, you can install it with `conda install vcstool`).

Once you have installed `vcstool`, that makes available the `vcs` command, go in the directory that contains the software repos (for the robotology-superbuild, go to `robotology-superbuild/src`) and run  : 
~~~
 vcs export --exact > used_git_repos.yaml
~~~

This will generate a `used_git_repos.yaml` file that contains the info of each used repo:
~~~
traversaro@IITICUBLAP102:~/robotology-superbuild/src$ head used_git_repos.yaml
repositories:
  icub-gazebo:
    type: git
    url: https://github.com/robotology/icub-gazebo.git
    version: 024bb76df3afa37ac534262daa78d5aaa36d8d25
  icub-models:
    type: git
    url: https://github.com/robotology/icub-models.git
    version: 67b4668ff8491d94d7269bc878319bf0bf78ce4c
~~~

Note that we used the `--exact` argument, otherwise the yaml would have contained the names of the current branches used by the repo, that is not useful for reproducibility. 

## firmware versions 

If you interested in the state of the software in the `icub-head` (or `pc104` for older robots), then you should consider that the behavior of the robot is also influenced by the specific **version of the firmware** loaded in the various low level boards of the iCub! To dump the state of that, follow the instructions in https://icub-tech-iit.github.io/documentation/icub_firmware, and run the `FirmwareUpdater.script.info.all.sh` script to collect this information. To be perfectly honest, I never tried the script itself, so more users with more experience with the firmware can probably add some info here (cc @violadelbono @marcoaccame @davidetome ).

**Reply by pattacini** [reply]
Great piece of information @traversaro !
We may want to add up a page publishing it in our official documentation.

cc @robotology/everyone

**Reply by traversaro** [reply]
>  We may want to add up a page publishing it in our official documentation.

That is a good idea! For now I just posted as a discussion because there are a few open points:
* Is the firmware `FirmwareUpdater.script.info.all.sh` script working? I did not tested it my self. 
* Capturing the dependencies of the robotology-superbuild repo is not really convenient if they are contained in several directories. https://github.com/robotology/robotology-superbuild/pull/556 should fix this.

**Reply by traversaro** [reply]
I updated the answer as now all the repos of the superbuild are cloned in `robotology-superbuild/src` .

**Reply by marcoaccame** [reply]
Hi @traversaro, 
> To dump the state of that, follow the instructions in http://wiki.icub.org/wiki/Firmware#Automatic_update_of_the_robot, and run the FirmwareUpdater.script.info.all.sh script to collect this information. To be perfectly honest, I never tried the script itself, so more users with more experience with the firmware can probably add some info here (cc @violadelbono @marcoaccame @davidetome ).

the script `FirmwareUpdater.script.info.all.sh` tells which FW versions the boards on the robot should have.
To query the actual state of the robot vs the target FW versions, one can use the `verify` scripts as described in [here](https://github.com/robotology/icub-firmware-build/blob/master/docs/FirmwareUpdater.readme.scripts.txt#L77-L79).
